File: desktop/src/Beam.Core/Security/ResumeHandshake.cs
using System.Security.Cryptography;

namespace Beam.Core.Security;

/// <summary>
/// Derivações do handshake de retomada — a reconexão automática que dispensa o
/// código de seis dígitos enquanto a confiança estiver válida.
/// </summary>
/// <remarks>
/// <para>
/// No pareamento por código, além da chave de sessão, ambos os lados derivam do
/// mesmo segredo ECDH uma <b>chave de retomada</b> de longa duração (info HKDF
/// distinta, portanto independente da chave de sessão). Ela é persistida dos dois
/// lados e expira 72h após a última conexão bem-sucedida.
/// </para>
/// <para>
/// Na reconexão, o celular envia <c>session.resume</c> com uma chave pública
/// X25519 efêmera nova e um nonce; o desktop responde com as suas. A chave de
/// sessão sai de <c>HKDF(ECDH ‖ chaveDeRetomada)</c>: o ECDH efêmero preserva o
/// sigilo futuro (uma retomada gravada hoje não é decifrável com a chave de
/// retomada vazada amanhã) e a chave de retomada autentica os dois lados. As
/// provas HMAC fecham o handshake em ambas as direções.
/// </para>
/// </remarks>
public static class ResumeHandshake
{
    /// <summary>Validade da confiança, renovada a cada conexão segura.</summary>
    public static readonly TimeSpan TrustLifetime = TimeSpan.FromHours(72);

    /// <summary>Tamanho dos nonces trocados no handshake de retomada.</summary>
    public const int NonceSizeBytes = 32;

    /// <summary>Info HKDF da chave de retomada (derivada no pareamento por código).</summary>
    public static readonly byte[] ResumeKeyInfo = "clipbridge-v1-resume"u8.ToArray();

    private static readonly byte[] EcdhInfo = "clipbridge-v1-resume-ecdh"u8.ToArray();
    private static readonly byte[] SessionInfo = "clipbridge-v1-resume-session"u8.ToArray();
    private static readonly byte[] DeviceIdInfo = "clipbridge-v1-device-id"u8.ToArray();
    private static readonly byte[] ServerProofLabel = "clipbridge-v1-resume-server"u8.ToArray();
    private static readonly byte[] ClientProofLabel = "clipbridge-v1-resume-client"u8.ToArray();

    /// <summary>
    /// Identificador público do vínculo, derivado da chave de retomada — os dois
    /// lados chegam ao mesmo valor sem trocá-lo, e ele não revela a chave.
    /// </summary>
    public static string ComputeDeviceId(ReadOnlySpan<byte> resumeKey) =>
        Convert.ToHexString(HMACSHA256.HashData(resumeKey, DeviceIdInfo).AsSpan(0, 8));

    /// <summary>Info HKDF do ECDH efêmero da retomada.</summary>
    public static ReadOnlySpan<byte> EcdhSessionInfo => EcdhInfo;

    /// <summary>Salt comum do handshake: os dois nonces, na ordem cliente‖servidor.</summary>
    public static byte[] BuildSalt(ReadOnlySpan<byte> clientNonce, ReadOnlySpan<byte> serverNonce)
    {
        var salt = new byte[clientNonce.Length + serverNonce.Length];
        clientNonce.CopyTo(salt);
        serverNonce.CopyTo(salt.AsSpan(clientNonce.Length));
        return salt;
    }

    /// <summary>
    /// Chave de sessão da retomada: o ECDH efêmero misturado com a chave de
    /// retomada. Sem a chave de retomada não se chega a ela, mesmo com o ECDH.
    /// </summary>
    public static byte[] DeriveSessionKey(
        ReadOnlySpan<byte> ecdhKey,
        ReadOnlySpan<byte> resumeKey,
        ReadOnlySpan<byte> salt)
    {
        var material = new byte[ecdhKey.Length + resumeKey.Length];
        try
        {
            ecdhKey.CopyTo(material);
            resumeKey.CopyTo(material.AsSpan(ecdhKey.Length));
            var sessionKey = new byte[SessionCipher.KeySizeBytes];
            HKDF.DeriveKey(HashAlgorithmName.SHA256, material, sessionKey, salt, SessionInfo);
            return sessionKey;
        }
        finally
        {
            CryptographicOperations.ZeroMemory(material);
        }
    }

    /// <summary>Prova de que o desktop conhece a chave de retomada.</summary>
    public static byte[] ServerProof(ReadOnlySpan<byte> resumeKey, ReadOnlySpan<byte> salt) =>
        Proof(resumeKey, ServerProofLabel, salt);

    /// <summary>Prova de que o celular conhece a chave de retomada.</summary>
    public static byte[] ClientProof(ReadOnlySpan<byte> resumeKey, ReadOnlySpan<byte> salt) =>
        Proof(resumeKey, ClientProofLabel, salt);

    private static byte[] Proof(ReadOnlySpan<byte> resumeKey, ReadOnlySpan<byte> label, ReadOnlySpan<byte> salt)
    {
        var message = new byte[label.Length + salt.Length];
        label.CopyTo(message);
        salt.CopyTo(message.AsSpan(label.Length));
        return HMACSHA256.HashData(resumeKey, message);
    }
}

File: mobile/app/src/main/java/com/esousa/beam/security/ResumeHandshake.kt
package com.esousa.beam.security

import org.bouncycastle.crypto.digests.SHA256Digest
import org.bouncycastle.crypto.generators.HKDFBytesGenerator
import org.bouncycastle.crypto.params.HKDFParameters
import java.nio.charset.StandardCharsets
import java.security.MessageDigest
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

/**
 * Derivações do handshake de retomada — espelho de `Beam.Core.Security.ResumeHandshake`.
 *
 * No pareamento por código, os dois lados derivam do mesmo segredo ECDH uma
 * chave de retomada de longa duração (info HKDF distinta da chave de sessão) e a
 * persistem. Na reconexão, a chave de sessão sai de `HKDF(ECDH efêmero ‖ chave de
 * retomada)`: o ECDH novo preserva o sigilo futuro, a chave de retomada autentica
 * os dois lados. Qualquer divergência de bytes aqui quebra a interoperabilidade
 * com o desktop.
 */
object ResumeHandshake {

    /** Validade da confiança, renovada a cada conexão segura. */
    const val TRUST_LIFETIME_MS = 72L * 60 * 60 * 1000

    const val NONCE_SIZE = 32

    /** Info HKDF da chave de retomada, derivada durante o pareamento por código. */
    val RESUME_KEY_INFO = "clipbridge-v1-resume".toByteArray(StandardCharsets.UTF_8)

    private val ECDH_INFO = "clipbridge-v1-resume-ecdh".toByteArray(StandardCharsets.UTF_8)
    private val SESSION_INFO = "clipbridge-v1-resume-session".toByteArray(StandardCharsets.UTF_8)
    private val DEVICE_ID_INFO = "clipbridge-v1-device-id".toByteArray(StandardCharsets.UTF_8)
    private val SERVER_PROOF_LABEL = "clipbridge-v1-resume-server".toByteArray(StandardCharsets.UTF_8)
    private val CLIENT_PROOF_LABEL = "clipbridge-v1-resume-client".toByteArray(StandardCharsets.UTF_8)

    /** Info HKDF do ECDH efêmero da retomada. */
    val ecdhInfo: ByteArray get() = ECDH_INFO

    /**
     * Identificador público do vínculo: os dois lados chegam ao mesmo valor sem
     * trocá-lo, e ele não revela a chave de retomada.
     */
    fun computeDeviceId(resumeKey: ByteArray): String =
        hmac(resumeKey, DEVICE_ID_INFO).take(8).joinToString("") { "%02X".format(it) }

    /** Salt comum do handshake: os dois nonces, na ordem cliente‖servidor. */
    fun buildSalt(clientNonce: ByteArray, serverNonce: ByteArray): ByteArray =
        clientNonce + serverNonce

    /** Chave de sessão da retomada: ECDH efêmero misturado com a chave de retomada. */
    fun deriveSessionKey(ecdhKey: ByteArray, resumeKey: ByteArray, salt: ByteArray): ByteArray {
        val material = ecdhKey + resumeKey
        try {
            return ByteArray(SessionCipher.KEY_SIZE).also { sessionKey ->
                HKDFBytesGenerator(SHA256Digest()).apply {
                    init(HKDFParameters(material, salt, SESSION_INFO))
                    generateBytes(sessionKey, 0, sessionKey.size)
                }
            }
        } finally {
            material.fill(0)
        }
    }

    /** Prova de que o desktop conhece a chave de retomada. */
    fun serverProof(resumeKey: ByteArray, salt: ByteArray): ByteArray =
        hmac(resumeKey, SERVER_PROOF_LABEL + salt)

    /** Prova de que este celular conhece a chave de retomada. */
    fun clientProof(resumeKey: ByteArray, salt: ByteArray): ByteArray =
        hmac(resumeKey, CLIENT_PROOF_LABEL + salt)

    /** Comparação em tempo constante — nunca vaza quanto do prefixo bateu. */
    fun matches(a: ByteArray, b: ByteArray): Boolean = MessageDigest.isEqual(a, b)

    private fun hmac(key: ByteArray, message: ByteArray): ByteArray =
        Mac.getInstance("HmacSHA256").run {
            init(SecretKeySpec(key, "HmacSHA256"))
            doFinal(message)
        }
}
